### PR TITLE
some features for the paper

### PR DESCRIPTION
**Pull request by o-**
1. Add a disassembler to produce ott compatible output that can be c/p into the tex
2. Untangle some optimizations to be able to show the changes in smaller increments.

### CODE REVIEW

File: transform.ml [hunk 1]
```diff
@@ -130,22 +139,17 @@ let optimize (opts : string list) (prog : program) : program option =
...
-    if opts_assumpt <> []
-    then combine_opt [
+    combine_opt [
         (as_opt_program Transform_assumption.insert_checkpoints);
         optimizer;
-        (as_opt_program (as_opt_function Transform_assumption.hoist_assumption));
         optimizer;
         Transform_assumption.remove_empty_osr;
-        optimizer_classic;
+        optimizer;
```
Comment (mhyee) [inline transform.ml]: Remind me again, why do we need `optimizer` repeatedly here?

**Review by mhyee — COMMENTED** [review]
It looks like you forgot to add some files (`ott.ml`?).

File: transform.ml (hunk 1 above)
Comment (o-) [reply transform.ml]: no good answer...

some transformations might enable other, so running multiple times makes more optimizations.

maybe we could instead have a cli argument for it...

**Comment by o-** [comment]
ok, now I added configurable optimizer pass number.
Also added the missing ott.ml file.
The code duplication is intentional since for now it would have been a bigger refactoring to make the disassembler configurable.

**Comment by o-** [comment]
the ott syntax is a bit strange because I need to differ between normal instruction rendering and aligned instruction rendering.

The normal syntax of a program is:
```
fun1 -> (x) ( version1 -> ( l1 -> return x ) ) ),
fun2 -> (y) ( versionx -> ( l1 -> return y ) ) )
```

If you want aligned instructions (ie. the instruction rendered as `l1 & : & return x` you can enclose the program in `P( ... )`, or a plain list of labeled instructions in `i( ... )`.

File: ott.ml [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        pr buf " read %s\n" var;
+        dump_next_instr ()
+      | Osr {label; cond; target={func; version; pos}; varmap; frame_maps} ->
+        print_label label;
+        let dump_var buf = function
+          | Osr_var (x, e)     -> pr buf "%s = %a" x dump_expr e
+        in
+        let dump_frame buf {cont_pos={func; version; pos}; cont_res; varmap} =
+          pr buf "(%s, %s, %s) [%s = $, %a]"
+              func version pos
```
Comment (mhyee) [inline ott.ml]: Nitpick: indentation should be 2 spaces not 4. (See the `pr buf ...` below.)

File: ott.ml [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  in
+  dump_instr buf 0 true
+
+let disassemble buf (prog : Instr.program) =
+  let dump_func buf {name; formals; body} =
+      let print_formal buf (Param x) = pr buf "%s " x in
+      let print_formals buf = (dump_comma_separated print_formal) buf formals in
+      Printf.bprintf buf "%s -> ( %t ) ( " name print_formals;
+      let dump_vers buf version =
+          pr buf "%s -> (\n" version.label;
```
Comment (mhyee) [inline ott.ml]: Nitpick: should be 2 space indentation.

File: sourir.ml [hunk 4]
```diff
@@ -13,6 +15,8 @@ let () =
     ("--quiet", Arg.Set quiet, "quiet");
     ("--autofix", Arg.Set autofix, "automatically normalize graph");
     ("--opt", Arg.String (fun s -> opts := String.split_on_char ',' s), "Enable optimizations");
+    ("--nopt", Arg.String (fun s -> opt_runs := int_of_string s), "Number of optimization runs");
```
Comment (mhyee) [inline sourir.ml]: Could you rename this to `opt-runs` or `num-opts`? `nopt` looks too much like `nop` or `no opt` to me.

File: ott.ml [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      | Comment str                     ->
+        dump_next_instr ()
+      end;
+    end
+  in
+  dump_instr buf 0 true
+
+let disassemble buf (prog : Instr.program) =
+  let dump_func buf {name; formals; body} =
+      let print_formal buf (Param x) = pr buf "%s " x in
```
Comment (mhyee) [inline ott.ml]: Nitpick: should be 2 space indentation.

**Review by mhyee — APPROVED** [review]

**Pull request merged**